### PR TITLE
J468 copy safe v2

### PR DESCRIPTION
**Pull request by rfree-d**
int_to_enum takes always int; assert overflowing. UT

### CODE REVIEW

File: src/test/stdplus_misc.cpp [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+	EXPECT_NO_THROW( { auto color = int_to_enum<t_color_components>(0);  _UNUSED(color); } );
+	EXPECT_NO_THROW( { auto color = int_to_enum<t_color_components>(2);  _UNUSED(color); } );
+	for (int i=0; i<=8; ++i) EXPECT_NO_THROW( { auto color = int_to_enum<t_color_components>(i);  _UNUSED(color); } );
+	for (int i=9; i<=300; ++i) EXPECT_THROW( { auto color = int_to_enum<t_color_components>(i);  _UNUSED(color); } , std::exception ); // no such base colors
+}
+
+// ===========================================================================================================
+// non-class enum (deprecated)
+
+enum t_gender { male=1, female=0 };
```
Comment (robertoleksy) [inline src/test/stdplus_misc.cpp]: feminists will be not happy

File: src/crypto/crypto.cpp [hunk 2]
```diff
@@ -311,8 +311,10 @@ c_crypto_system::t_symkey c_stream::calculate_KCT
 
 	for (size_t sys=0; sys<self.m_pub.get_count_of_systems(); ++sys) { // all key crypto systems
 		// for given crypto system:
-
-		auto sys_enum = int_to_enum<t_crypto_system_type>(sys); // enum of this crypto system
+		t_crypto_system_type sys_enum;
+		try {
+			sys_enum = int_to_enum<t_crypto_system_type>(sys,true); // enum of this crypto system
+		} catch(expected_not_found) { continue ; } // not assigned enum type, skip
```
Comment (rfree-d) [inline src/crypto/crypto.cpp]: ok (and updated doc/hacking)

File: src/crypto/multikeys.cpp [hunk 3]
```diff
@@ -264,17 +268,20 @@ c_multisign c_multikeys_PRV::multi_sign(const string &msg) {
 	// all key crypto systems
 	for (size_t sys=0; sys<get_count_of_systems(); ++sys) {
 
-		auto crypto_type = int_to_enum<t_crypto_system_type>(sys); // enum of this crypto syste
+		t_crypto_system_type sys_enum;
+		try {
+			sys_enum = int_to_enum<t_crypto_system_type>(sys,true); // enum of this crypto system
+		} catch(expected_not_found) { continue ; } // not assigned enum type, skip
```
Comment (rfree-d) [inline src/crypto/multikeys.cpp]: ditto

File: src/crypto/multikeys.cpp [hunk 4]
```diff
@@ -371,14 +378,18 @@ void c_multikeys_PAIR::generate(t_crypto_system_count cryptolists_count, bool wi
 	_info("Generating from cryptolists_count");
 	for (size_t sys=0; sys<cryptolists_count.size(); ++sys) { // all key crypto systems
 		// for given crypto system:
-		auto sys_id = int_to_enum<t_crypto_system_type>(sys); // ID of this crypto system
+		t_crypto_system_type sys_enum;
+		try {
+			sys_enum = int_to_enum<t_crypto_system_type>(sys,true); // enum of this crypto system
+		} catch(expected_not_found) { continue ; } // not assigned enum type, skip
```
Comment (rfree-d) [inline src/crypto/multikeys.cpp]: ditto

File: src/stdplus/misc.hpp [hunk 5]
```diff
@@ -13,12 +13,40 @@ and stdplus/misc are assorted parts of that
...
+
+/// Throw this if some element was not found, but this is normal situation.
+/// E.g. when public-key was not yet created at all (to differentiate
+/// from case where public-key was found but can not be loaded due to format errors etc)
+class expected_not_found : public stdplus::expected_exception {
+	public:
+		const char* what() const noexcept override;
+};
+
+template<typename T> constexpr bool templated_always_false() { return false; }
```
Comment (rfree-d) [inline src/stdplus/misc.hpp]: documented why it was needed, and switched to the idea you recommended with ,FALSE template

File: src/stdplus/misc.hpp [hunk 6]
```diff
@@ -59,69 +95,125 @@ T* unique_cast_ptr(std::unique_ptr<U> & u) {
...
-template<typename T> constexpr bool templated_always_false() { return false; }
+/// @return the object pointed to by given iterator, in form of void*;
+/// @warning the iterator iter must be an AddressibleIterator
+template <typename T>
+const void * iterator_to_voidptr(T iter
+	,typename std::enable_if< // Concept
+		std::is_same<  typename std::iterator_traits<T>::iterator_category , std::random_access_iterator_tag	 >
+	::value >::type * = 0)
+{
+	typedef typename std::iterator_traits<T>::value_type value_type;
```
Comment (rfree-d) [inline src/stdplus/misc.hpp]: ok

File: src/stdplus/misc.hpp [hunk 7]
```diff
@@ -59,69 +95,125 @@ T* unique_cast_ptr(std::unique_ptr<U> & u) {
...
+
+/// @return string representing debug information about given iterator
+/// @warning the iterator iter must be an AddressibleIterator
+template <typename T>
+std::string to_debug(const T & X, t_debug_style style=e_debug_style_object
+	,typename std::enable_if< // Concept, needed to choose from other functions with same name
+		std::is_same<  typename std::iterator_traits<T>::iterator_category , std::random_access_iterator_tag	 >
+	::value >::type * = 0)
+{
+	(void)(style); // unused
```
Comment (rfree-d) [inline src/stdplus/misc.hpp]: was not possible here, moved it to other header-file, now works

File: src/stdplus/misc.hpp [hunk 8]
```diff
@@ -29,14 +34,19 @@ std::string STR(const T & obj) { ///< stringify any object
...
 
-template <typename TE, typename TI>
-TE int_to_enum(TI i) { ///< convert integer into enum of given type, throw if this value is not represented in target enum
+template <typename TE>
+TE int_to_enum(int i) { ///< convert integer into enum of given type, throw if this value is not represented in target enum
+	// TODO assert that the type TE is an defined enum here?
+	static_assert( std::is_enum<TE>::value , "Must be an enum type (defined enum)");
+
+	// assert that the type i will over/under flow (e.g. wrap back to some valid value causing undetected error)
+	_check_input( static_cast<long long int>(i)  <= static_cast<long long int>( std::numeric_limits< typename std::underlying_type<TE>::type >::max() ) ); // TODO it's afe compare - upcast both types to long long... make tool for this idiom?
```
Comment (rfree-d) [inline src/stdplus/misc.hpp]: done it correctly. Actually we also need to properly cast to underlying type and check for overflows.
Using boost numeric_cast.

File: src/test/stdplus_misc.cpp (hunk 1 above)
Comment (rfree-d) [inline src/test/stdplus_misc.cpp]: ok, ladiest first :woman:

**Pull request merged**